rust/lance/src/index/vector/pq.rs

```diff
@@ -263,5 +268,246 @@ impl VectorIndex for PQIndex {
+) -> Result<Arc<dyn ProductQuantizer>> {
+    if let Some(codebook) = &params.codebook {
+        let mt = if metric_type == MetricType::Cosine {
+            info!("Normalize training data for PQ training: Cosine");
+            MetricType::L2
+        } else {
+            metric_type
+        };
+
+        return match codebook.data_type() {
+            DataType::Float16 => Ok(Arc::new(ProductQuantizerImpl::<Float16Type>::new(
+                params.num_sub_vectors,
+                params.num_bits as u32,
+                dim,
+                Arc::new(codebook.as_primitive().clone()),
+                mt,
+            ))),
+            DataType::Float32 => Ok(Arc::new(ProductQuantizerImpl::<Float32Type>::new(
+                params.num_sub_vectors,
+                params.num_bits as u32,
+                dim,
+                Arc::new(codebook.as_primitive().clone()),
+                mt,
+            ))),
+            DataType::Float64 => Ok(Arc::new(ProductQuantizerImpl::<Float64Type>::new(
+                params.num_sub_vectors,
+                params.num_bits as u32,
+                dim,
+                Arc::new(codebook.as_primitive().clone()),
+                mt,
+            ))),
+            _ => {
+                return Err(Error::Index {
+                    message: format!("Wrong codebook data type: {:?}", codebook.data_type()),
+                    location: location!(),
+                });
+            }
+        };
+    }
+    info!(
+        "Start to train PQ code: PQ{}, bits={}",
+        params.num_sub_vectors, params.num_bits
+    );
+    let expected_sample_size =
+        lance_index::vector::pq::num_centroids(params.num_bits as u32) * params.sample_rate;
+    info!(
+        "Loading training data for PQ. Sample size: {}",
+        expected_sample_size
+    );
+    let start = std::time::Instant::now();
+    let mut training_data =
+        maybe_sample_training_data(dataset, column, expected_sample_size).await?;
+    info!(
+        "Finished loading training data in {:02} seconds",
+        start.elapsed().as_secs_f32()
+    );
+
+    info!(
+        "starting to compute partitions for PQ training, sample size: {}",
+        training_data.value_length()
+    );
+
+    if metric_type == MetricType::Cosine {
+        info!("Normalize training data for PQ training: Cosine");
+        training_data = normalize_fsl(&training_data)?;
+    }
+
+    println!("PQ Training, ivf: {:?} ", ivf);
+
+    let training_data = if let Some(ivf) = ivf {
+        // Compute residual for PQ training.
+        //
+        // TODO: consolidate IVF models to `lance_index`.
+        let ivf2 = lance_index::vector::ivf::new_ivf(
+            ivf.centroids.values(),
+            ivf.dimension(),
+            MetricType::L2,
+            vec![],
+            None,
+        )?;
+        span!(Level::INFO, "compute residual for PQ training")
+            .in_scope(|| ivf2.compute_residual(&training_data, None))
+            .await?
```

Comment:
I think we still need to make sure we don't residulize for dot.

So, make something like `pq_params.should_residulize`?

Reply:
So the design lets the caller decide whether to run residual or not. If provided IVF, it runs residual PQ.

So PQ is orthogonal to distance type as much as possible.

Reply:
but we always residualize currently.

Reply:
need a flag for dot distance to not residualize

Reply:
Ok, checked it on the caller